packages/flame/lib/src/components/component.dart

```diff
@@ -74,7 +76,16 @@ class Component {
   ///
   /// It will be checked once per component per tick, and if it is true,
   /// FlameGame will remove it.
-  bool shouldRemove = false;
+  @nonVirtual
+  bool get shouldRemove => _state == LifecycleState.removing;
+
+  /// Legacy component removal mechanism.
+  @nonVirtual
+  @Deprecated('Use `removeFromParent()` instead. This will be removed in 1.1.0')
```

Comment:
If we remove it in `1.1.0` it will never have a release in-between where it is just deprecated, was that intentional?

Reply:
Haven't you argued for exactly this approach in #1316? Ever since then this is how we've been handling deprecation (searching for `@Deprecated` in the code shows that in all instances we're announcing deprecation in 1.1.0).

It's not too late to change the policy though and revisit our current deprecations.

Reply:
I'll have a check with the team on the meeting tomorrow, but I think for these cases that are easy to put a deprecation marker on and at the same time have the new functionality it would be nice to have it out for one release.

Reply:
Since this is a setter now, I don't see the need to deprecate and eventually remove it, we could keep it.

I see value having this attribute it allows things like

'''dart
comp.shouldRemove = position.x < 0
'''

Reply:
> I don't see the need to deprecate and eventually remove it, we could keep it.

I feel that a setter that works as a method is a bit of an anti-pattern. If you think that this is a useful method to have (and it kinda does look like a useful thing), then we could replace the setter with an actual method such as
```dart
void removeIf(bool shouldRemove) {
  if (shouldRemove) removeFromParent();
}
```

Reply:
I like the removeIf method, but idk, this is a breaking change that we could avoid, so I will have to put some thought on this

Reply:
I see it like this:

`shouldRemove`
Pro: doesn't have to be called manually in `update`.
Con: adds another way to remove.

I don't see the use for `removeIf`, the `if` logic could just be directly in the update-loop in that case?

I think I'm leaning towards the option of deprecating it to minimize the amount of ways to do removals, but I don't have any strong opinion on the matter.

Reply:
We still haven't decided whether `shouldRemove` should be deprecated or not, we'll probably have to vote about it.
But meanwhile, this method does not do the same thing as it did before this PR. Previously you could for example override the getter with logic for removal and you could also set it to true later in the same tick even if it had been set to false earlier.

Reply:
@spydon The use case that you're describing, even though it's very rare, is still possible to implement with this PR. All you need is to create your own flag `toRemove`, then set it to true/false as many times as needed, then at the end of the `update()` method check the flag and call `.removeFromParent()` if it's true. That would do exactly what the current functionality is doing, but explicitly.

In general, I feel explicit deletion is better than implicit because 
 1) it's more clear what is going on and when the deletion happens;
 2) fewer things to document for us, fewer things for the user to get confused about;
 3) components that don't need to be removed don't pay the cost of querying the flag on every game tick.

Reply:
I'm also on the side of deletion. But we should still try to avoid a direct breaking change here if we can.

> The use case that you're describing, even though it's very rare, is still possible to implement with this PR

It's not about whether you can or can't do it, it's that it changes the current behavior. So right now it will be a breaking change even though it is just deprecated. Might not matter though, I mean we just have to call it.

> The use case that you're describing, even though it's very rare.

It's not that rare sadly (overriding the getter), I've seen several projects with that logic embedded in the getter (I think it must have been in some external tutorial at some point).

```dart
@override
bool get shouldRemove => "not within screen" or "life below 0";
```

Reply:
Ah, I see your point. My "very rare" comment was about deleting something and then deciding to undelete it in the very same tick. But what you were thinking is overriding the `shouldRemove` getter. That would indeed be more common, so we should probably call it out explicitly.

Right now the getter is marked as `@nonVirtual`, so the users will at least be aware that something in their code needs to change.

Reply:
It was decided in a very even vote today that `shouldRemove` shouldn't be deprecated.

Reply:
@spydon Is there any specific reason? The cost of `shouldRemove` is that every component has to check this flag on every single game tick.

Reply:
I voted for removing it, but the argument for keeping it was that we should minimize breaking changes. Just checking a bool shouldn't be a performance concern though, if you can show that it is I'm sure they'd change their minds, but I would also be very worried about Dart as a language if it was true. 😅

Reply:
It's just not a very nice property. For example, if you override its getter then it'll become inconsistent with the setter. Maybe we should at least override the setter then? It's equivalent to the `remove` method anyway.

Reply:
> Maybe we should at least override the setter then?

Sounds good to me, how were you thinking that the override would look?

Reply:
I meant, deprecate the setter. Because having both the setter and an overridable getter makes no sense to me.

Reply:
Ah, I'll ask the team.

Reply:
Can't we just set `@nonVirtual` on the setter?
The team didn't think it made much sense to only keep the getter since that would only be useful when overridden then.

Reply:
In order to understand whether any kind of API is good or not, we need to think about how we're going to document it. And by "document" I don't mean to write a one-liner that just repeats the name of the function like "set whether this component should be removed" -- that's not a proper documentation. A great documentation follows the principles outlined here: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#writing-prompts-for-good-documentation

So, a good documentation for setter will look like this:
```
/// Setting this property to true will remove the component soon (before the next 
/// game cycle), thus it is equivalent to `component.removeFromParent()`. However,
/// if you override `.shouldRemove` getter, then there is no guarantee that setting
/// this property to true will have any effect.
```

And when you spell it out, it kinda becomes clear that it's not a very good API. Here, by "not good" I mean that it (1) has no real purpose (the same functionality can be achieved via more straightforward call to `.remove`), (2) does not work reliably.

But ok, if we want to keep it, then let's keep it.

Reply:
I agree that it is not optimal, but the vote is done. It will be reliable though, since they'll be marked as `@nonVirtual` and can't be overridden.

Reply:
Alright, i marked the getter and the setter `.shouldRemove` as `@nonVirtual`, and removed deprecation annotation.